src/relation_engine_server/server.py

```diff
@@ -25,137 +31,90 @@ def root():
-    """
-    Fetch the JSON for a single schema.
-    Auth: public
-    """
-    return flask.jsonify(spec_loader.get_schema(name))
-
-
-@app.route('/api/views/<name>', methods=['GET'])
-def show_view(name):
-    """
-    Fetch the AQL for a single view.
-    Auth: public
-    """
-    return flask.Response(spec_loader.get_view(name), mimetype='text/plain')
-
-
-@app.route('/api/update_specs', methods=['GET'])
-def refresh_specs():
-    """
-    Manually check for updates, download spec releases, and init new collections.
-    Auth: admin
-    """
-    auth.require_auth_token(['RE_ADMIN'])
-    init_collections = 'init_collections' in flask.request.args
-    release_url = flask.request.args.get('release_url')
-    pull_spec.download_specs(init_collections, release_url)
-    return flask.jsonify({'status': 'updated'})
+    }
+    return _json_resp(body)
 
 
-@app.route('/api/documents', methods=['PUT'])
-def save_documents():
+@app.route('/api/<path:path>', methods=['GET', 'PUT', 'POST', 'DELETE'])
+def api_call(path):
     """
-    Create, update, or replace many documents in a batch.
-    Auth: admin
+    Handle an api request, dispatching it to the appropriate versioned module.
+
+    Versioning system:
+    - Every API version is a discrete python module that contains an 'endpoints' dictionary.
+    - New versions don't need to redefine previous endpoints.
+        - If an endpoint is not defined, we fall back to a previous version.
+    - New versions can overwrite existing methods and add new ones.
+    - Versions are simple incrementing integers. We only need a new version for breaking changes.
+    - New modules can deprecate paths by putting them under the 'deprecations' key
+    Note that endpoints cannot be removed with new versions, only overwritten or added.
     """
-    auth.require_auth_token(['RE_ADMIN'])
-    collection_name = flask.request.args['collection']
-    query = {'collection': collection_name, 'type': 'documents'}
-    if flask.request.args.get('display_errors'):
-        # Display an array of error messages
-        query['details'] = 'true'
-    if flask.request.args.get('on_duplicate'):
-        query['onDuplicate'] = flask.request.args['on_duplicate']
-    if flask.request.args.get('overwrite'):
-        query['overwrite'] = 'true'
-    resp_text = bulk_import.bulk_import(query)
-    return resp_text
+    path_parts = path.split('/')
+    version_int = _get_version(path_parts[0])
+    # Get the path and version number
+    api_path = '/'.join(path_parts[1:])
+    # Flag for whether the endpoint we find is deprecated
+    deprecated = False
+    # Find our method in the various versioned modules
+    # If it is not present in a later version, fall back to a previous version
+    # Iterates by starting at (version-1), stopping at 0, and stepping backwards
+    # Note: the mypy type checker has difficulties with the path_funcs dicts, so we ignore type checking below
+    for ver in range(version_int - 1, -1, -1):
```

Comment:
Hmm, I think this is to permit code reuse without boilerplate but I don't think I like the idea of automatically falling back to an older version of the api to find missing function. For example, we used to call the method "update_specs", "refresh_specs". With this versioning, I might forget that rename and accidentally call "refresh_specs" even though I intended to call "update_specs" with it's modern paramenters and behaviors, I just botched the name. I should get a  view_does_not_exist error not the older function if I make that call. Effectively this structure means that methods can be depreciated or redefined but never removed from future versions of an API

I also foresee an issue with maintenance. If old endpoints that are pulled though are explicitly specified in each version of the API, you can find the exact implementation for any API easily. with this, you need to look though the API files in reverse order until you find a method that matches just like this function.